sound/soc/sof/intel/hda.c

```diff
@@ -958,6 +958,27 @@ int hda_dsp_remove(struct snd_sof_dev *sdev)
 	return 0;
 }
 
+int hda_dsp_shutdown(struct snd_sof_dev *sdev)
+{
+	const struct sof_dsp_power_state target_state = {
+		.state = SOF_DSP_PM_D0,
+		.substate = SOF_HDA_DSP_PM_D0I0,
+	};
+	int ret;
+
+	if (sdev->dsp_power_state.state == SOF_DSP_PM_D0 &&
+	    sdev->dsp_power_state.substate == SOF_HDA_DSP_PM_D0I3) {
+		/* ensure the DSP is in D0 before shutting it down */
+		ret = snd_sof_dsp_set_power_state(sdev, &target_state);
+		if (ret < 0) {
+			dev_err(sdev->dev, "error at pulling DSP to D0 %d\n", ret);
```

Comment:
this doesn't seem very useful, the PCI core already resumes the device so it *IS* in DO

````c
static void pci_device_shutdown(struct device *dev)
{
    struct pci_dev *pci_dev = to_pci_dev(dev);
    struct pci_driver *drv = pci_dev->driver;

 

    pm_runtime_resume(dev);
````

Reply:
> this doesn't seem very useful, the PCI core already resumes the device so it _IS_ in DO
> 
> ```c
> static void pci_device_shutdown(struct device *dev)
> {
>     struct pci_dev *pci_dev = to_pci_dev(dev);
>     struct pci_driver *drv = pci_dev->driver;
> 
>  
> 
>     pm_runtime_resume(dev);
> ```

We are not in runtime suspended but D0I3 when WoV is active, so no resume from PCI core neither. 

To be simple, when only WoV is active, DSP is put to D0I3, we need to switch it back to D0 before performing system reboot, otherwise, there might be reboot failure for DSP.